independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java

```diff
@@ -246,35 +247,45 @@ void notify(T event, ObserverExceptionHandler exceptionHandler, boolean async) {
+                                if (isTxObserver(om)) {
+                                    deferredEvents.add(new DeferredEventNotification<>(om, eventContext,
+                                            Status.valueOf(om.getTransactionPhase())));
+                                }
                             }
-                        }
 
-                        Synchronization sync = new ArcSynchronization(deferredEvents);
-                        TransactionSynchronizationRegistry registry = registryInstance.get();
-                        try {
-                            registry.registerInterposedSynchronization(sync);
-                            // registration succeeded, notify all non-tx observers synchronously
-                            predicate = predicate.and(this::isNotTxObserver);
-                        } catch (Exception e) {
-                            if (e.getCause() instanceof RollbackException || e.getCause() instanceof IllegalStateException) {
-                                // registration failed, AFTER_SUCCESS OMs are accordingly to CDI spec left out
-                                predicate = predicate.and(this::isNotAfterSuccess);
+                            Synchronization sync = new ArcSynchronization(deferredEvents);
+                            TransactionManager txManager = transactionManagerInstance.get();
+                            try {
+                                // NOTE - We are using standard synchronization on purpose as that seems more
+                                // fitting than interposed sync. Either way will have some use-cases that won't work.
+                                // See for instance discussions on https://github.com/eclipse-ee4j/cdi/issues/467
+                                txManager.getTransaction().registerSynchronization(sync);
+                                // registration succeeded, notify all non-tx observers synchronously
+                                predicate = predicate.and(this::isNotTxObserver);
+                            } catch (Exception e) {
+                                if (e.getCause() instanceof RollbackException
+                                        || e.getCause() instanceof IllegalStateException
+                                        || e.getCause() instanceof SystemException) {
+                                    // registration failed, AFTER_SUCCESS OMs are accordingly to CDI spec left out
+                                    predicate = predicate.and(this::isNotAfterSuccess);
+                                }
                             }
                         }
+                    } catch (SystemException e) {
+                        // In theory, this can be thrown by TransactionManager#getStatus() at which point we cannot even
+                        // determine if we should register some synchronization, therefore, no action is taken
```

Comment:
No entirely sure about this, but this seemed like the most reasonable approach to me. Comments are welcome :)

Reply:
Could not be added some debug logging that would print the exception with stacktrace, to be able to find the root cause on investigation?

Reply:
Good point, can add that tomorrow.

Reply:
Fixed this, thanks for pointing it out.